### DISCUSSION TITLE
Rollups Factories

### DISCUSSION THREAD

**Discussion by augustoteixeira** [cartesi/cips#13] [General]
Hi everyone,

I am proposing a design pattern in which all rollups would be created by a single Factory smart contract (https://github.com/cartesi/cips/blob/factory/cips/cip-6.md).

Any opinions on the design?

**Comment by tuler** [comment]
Not sure how to comment on the CIP. Will do here.

I like it.
Another advantage is to have an event emitted by the Factory which can be used to kick-off dynamic off-chain components for each new DApp. It can be:

- thegraph processing
- validator nodes
- or any component we decide to be multi-dapp

I don't see the reason to store the list of the DApps onchain. Isn't this going to increase cost? At what benefit?

**Reply by guidanoli** [reply]
I, too, think storing IDs and addresses on-chain would only be necessary if it were ever going to be queried by other smart contracts, because otherwise this information could be easily retrieved off-chain from the event log.

**Reply by tuler** [reply]
Even if other smart contracts call the factory, they can get the address of the newly create dapp as the return value of the create method, right? And store wherever they want.

**Reply by guidanoli** [reply]
Exactly, I think that's great for modularity. I don't think the factory should store an ID, because then each factory would have its own counter. Instead, and that is something I showed @felipeargento recently, I think there should be a super-factory that manages factories of all different versions and this one would be able to nicely assign serial IDs to applications.

**Comment by augustoteixeira** [comment]
Thanks!

I have added your comments to the CIP itself. I don't know how to proceed with this either :P

**Comment by guidanoli** [comment]
Nice job, Augusto!
You're right in saying that the Factory must:

1. Create a Diamond
2. Cut the facets into the Diamond

Notice that the call to `diamondCut` on step 2 must:

2. 
    a. Assign each function selector to the appropriate facet
    b. Initialize the facets via the provided initialization routine

Notice also that step 2.b should already create the _Setup Input_, as specified on _CIP-3_:

https://github.com/cartesi/cips/blob/611538bd343d4183378151b94d95dc66d6330544/cips/cip-3.md#L56

But, in _CIP-6_, you're saying that the Factory sends the _Setup Input_ to the Rollups contract.

https://github.com/cartesi/cips/blob/de631ef2143ba21681cd85a8d6a6e876dfb7a002/cips/cip-6.md#L42

https://github.com/cartesi/cips/blob/de631ef2143ba21681cd85a8d6a6e876dfb7a002/cips/cip-6.md#L59

This line can be misinterpreted as the Factory being the one adding the input, which is not the same thing.
We want the _Setup Input_ to be added by the Diamond itself, so that the off-chain machine knows the address of the DApp contract by the sender of the first input (which is the _Setup Input_).

**Observation:** I do agree, however, that the Factory indirectly ensures the first input to be the _Setup Input_, because it provides an initialization routine trusted by Cartesi, but I think this indirection must be made explicit in the _CIP-6_ as not to confound the reader into thinking it is the one who adds the _Setup Input_.

**Comment by augustoteixeira** [comment]
Thanks for the contributions, @guidanoli!

I have decided to remove any mention of Setup Inputs to this CIP. It is just confusing, since there was already a method to guarantee that the Setup Input was the first one.

Could you please make a PR emphasizing the exact steps that the factory will have to perform in order to properly deploy a diamond (cuts, facets, linking...)?

**Comment by guidanoli** [comment]
@augustoteixeira I'll be glad to do so. :-)

**Comment by tuler** [comment]
IMO the "setup input" looks like a hack that would be better designed if we had a factory.
The template hash would be part of the rollups "announcement" by the factory, "here is a new rollups and its template hash is this".

**Comment by guidanoli** [comment]
Also, I don't know if we need to emit two events:

* one emitted by the `Factory` with: constructor arguments, `RollupsId`, and DApp address
* one emitted by the DApp with: constructor arguments

The first one provides more information, so maybe we can remove the second one?
Or maybe remove constructor arguments from the event emitted by the `Factory`?

**Comment by guidanoli** [comment]
Wouldn't it be a problem for DApps created by different factories to have the same ID?
Likewise, if a DApp wants to upgrade, say, from 0.3.0 to 0.4.0, wouldn't it be preferrable that it had the same ID?
These issues can be solved if there is a "Version Manager" contract handling the assignment of IDs to new applications, and application upgrades to newer versions. Something like the following diagram:

```mermaid
graph TD
    VM[Rollups Version Manager] --> F3[Rollups Factory v0.3.0]
    VM --> F4[Rollups Factory v0.4.0]
    VM --> F5[Rollups Factory v0.5.0]
    F3 --> DApp1
    F3 --> DApp2
    F4 --> DApp3
```

Then, the owner of `DApp1` could call `upgrade`, which would make the necessary changes (like make the diamond cut with the appropriate initialization routine, add some input to tell the machine to upgrade some software, etc) and change the version of the application on the "Version Manager". Then, the diagram would look like this:

```mermaid
graph TD
    VM[Rollups Version Manager] --> F3[Rollups Factory v0.3.0]
    VM --> F4[Rollups Factory v0.4.0]
    VM --> F5[Rollups Factory v0.5.0]
    F4 --> DApp1
    F3 --> DApp2
    F4 --> DApp3
```

**Reply by tuler** [reply]
I still don't understand why we need an ID.

**Reply by guidanoli** [reply]
I believe it is because of the size of Ethereum address being too big to fit into a Chain ID.

**Reply by tuler** [reply]
What do you mean by `Chain ID`?

**Reply by guidanoli** [reply]
When signing a transaction to the blockchain, one needs to specify the Chain ID of the network it should be committed to. This is done to prevent attackers from reusing the same transaction across different blockchains.
When signing a transaction with MetaMask [[1]](https://docs.metamask.io/guide/sending-transactions.html), the extension automatically fills this Chain ID field, depending on which network you're currently connected to.
Talking to @augustoteixeira, it seems that it won't be necessary that each DApp has its own ID, and we'll be using a single Cartesi Chain ID for every DApp.

**Reply by guidanoli** [reply]
So, I think we can dismiss this design. :-)
